src/hotspot/share/logging/logStream.hpp

```diff
@@ -45,68 +51,218 @@ class LogStream : public outputStream {
   };
-  LineBuffer  _current_line;
-  LogTargetHandle _log_handle;
 
-  // Prevent operator new for LogStream.
-  static void* operator new (size_t);
-  static void* operator new[] (size_t);
+private:
+  LineBuffer _current_line;
+
+protected:
+  BackingLog _backing_log;
+
+public:
+  explicit LogStreamImpl(BackingLog bl)
+    : _backing_log(bl) {};
+
+  virtual ~LogStreamImpl() {
+    if (!_current_line.is_empty()) {
+      _backing_log.print("%s", _current_line.buffer());
+      _current_line.reset();
+    }
+  }
+
+  bool is_enabled() {
+    return _backing_log.is_enabled();
+  }
+
+  void write(const char* s, size_t len) override {
+    if (len > 0 && s[len - 1] == '\n') {
+      _current_line.append(s, len - 1); // omit the newline.
+      _backing_log.print("%s", _current_line.buffer());
+      _current_line.reset();
+    } else {
+      _current_line.append(s, len);
+    }
+    update_position(s, len);
+  }
+};
+
+template<typename T>
+LogStreamImpl<T>::LineBuffer::LineBuffer()
+  : _buf(_smallbuf), _cap(sizeof(_smallbuf)), _pos(0) {
+  _buf[0] = '\0';
+}
+
+template<typename T>
+LogStreamImpl<T>::LineBuffer::~LineBuffer() {
+  assert(_pos == 0, "still outstanding bytes in the line buffer");
+  if (_buf != _smallbuf) {
+    os::free(_buf);
+  }
+}
+
+// try_ensure_cap tries to enlarge the capacity of the internal buffer
+// to the given atleast value. May fail if either OOM happens or atleast
+// is larger than a reasonable max of 1 M. Caller must not assume
+// capacity without checking.
+template<typename T>
+void LogStreamImpl<T>::LineBuffer::try_ensure_cap(size_t atleast) {
+  assert(_cap >= sizeof(_smallbuf), "sanity");
+  if (_cap < atleast) {
+    // Cap out at a reasonable max to prevent runaway leaks.
+    const size_t reasonable_max = 1 * M;
+    assert(_cap <= reasonable_max, "sanity");
+    if (_cap == reasonable_max) {
+      return;
+    }
+
+    const size_t additional_expansion = 256;
+    size_t newcap = align_up(atleast + additional_expansion, additional_expansion);
+    if (newcap > reasonable_max) {
+      log_info(logging)("Suspiciously long log line: \"%.100s%s",
+                        _buf, (_pos >= 100 ? "..." : ""));
+      newcap = reasonable_max;
+    }
+
+    char* const newbuf = (char*)os::malloc(newcap, mtLogging);
+    if (newbuf == NULL) { // OOM. Leave object unchanged.
+      return;
+    }
+    if (_pos > 0) { // preserve old content
+      memcpy(newbuf, _buf, _pos + 1); // ..including trailing zero
+    }
+    if (_buf != _smallbuf) {
+      os::free(_buf);
+    }
+    _buf = newbuf;
+    _cap = newcap;
+  }
+  assert(_cap >= atleast, "sanity");
+}
+
+template<typename T>
+void LogStreamImpl<T>::LineBuffer::append(const char* s, size_t len) {
+  assert(_buf[_pos] == '\0', "sanity");
+  assert(_pos < _cap, "sanity");
+  const size_t minimum_capacity_needed = _pos + len + 1;
+  try_ensure_cap(minimum_capacity_needed);
+  // try_ensure_cap may not have enlarged the capacity to the full requested
```

Comment:
Would it be worth having `try_ensure_cap` return a boolean to indicate whether it was able to  expand to the size requested?

Reply:
I did this and also took the time to refactor `try_ensure_cap`, I think it reads more easily now.